boa/src/syntax/lexer/number.rs

```diff
@@ -253,34 +294,54 @@ impl<R> Tokenizer<R> for NumberLiteral {
             }
         }
 
-        // Consume digits until a non-digit character is encountered or all the characters are consumed.
-        cursor.take_while_ascii_pred(&mut buf, &|c: char| c.is_digit(kind.base()))?;
+        let next = if self.init == b'.' {
+            Some(b'.')
+        } else {
+            // Consume digits and separators until a non-digit non-separator
+            // character is encountered or all the characters are consumed.
+            take_integer(&mut buf, cursor, &kind, !legacy_octal)?;
+            cursor.peek()?
+        };
 
         // The non-digit character could be:
         // 'n' To indicate a BigIntLiteralSuffix.
-        // '.' To indicate a decimal seperator.
+        // '.' To indicate a decimal separator.
         // 'e' | 'E' To indicate an ExponentPart.
-        match cursor.peek()? {
+        match next {
             Some(b'n') => {
                 // DecimalBigIntegerLiteral
                 // Lexing finished.
-
                 // Consume the n
+                if legacy_octal {
+                    return Err(Error::syntax(
+                        "n suffix not allowed in octal representation",
+                        cursor.pos(),
+                    ));
+                }
                 cursor.next_byte()?.expect("n character vanished");
 
                 kind = kind.to_bigint();
             }
             Some(b'.') => {
                 if kind.base() == 10 {
-                    // Only base 10 numbers can have a decimal seperator.
+                    // Only base 10 numbers can have a decimal separator.
                     // Number literal lexing finished if a . is found for a number in a different base.
-
-                    cursor.next_byte()?.expect(". token vanished");
-                    buf.push(b'.'); // Consume the .
+                    if self.init != b'.' {
+                        cursor.next_byte()?.expect(". token vanished");
```

Comment:
```suggestion
                        cursor.next_byte()?.expect("'.' token vanished");
```
It's clearer to put the token in `'` (single quote) so we don't confuse it with punctuation from the text.